lib/DxcSupport/HLSLOptions.cpp

```diff
@@ -49,6 +49,10 @@ namespace {
 
 static HlslOptTable *g_HlslOptTable;
 
+#ifndef _WIN32
```

Comment:
Awesome stuff! TIL!

To make sure I understand this correctly: this hides the enclosed symbols from the `libDxcSupport.so` so that on Linux, you only see one copy of the enclosed symbols, which are statically linked into the dxc binary?

But we are still having two copies of static globals and they are initialized separately by the shared lib and binary on Windows/macOS? Is that a problem?

Reply:
I misstated the case in my explanation in the issue. There are and will always be two sets of static globals because the same object file is included in both the dxc binary and the dynamic library. The only difference is in which functions get called where on the various operating systems. If they always call into the same place like on linux, that's fine. If they call into one place from the library and another from the binary, I think that's weird, but that's what windows does and it's what apple can most easily be made to do. This way every platform does things closer to the same way. Additionally, all but the explicitly exported symbols are effectively hidden like this on Windows. I tried that too, but Apple had some weird issues, so I scaled it back.

Reply:
To be perfectly clear, when you said `libDxcSupport.so`, I assumed you meant `libdxcompiler.so` `dxcsupport` is compiled into static library and linked into `dxc` and `libdxcompiler.so. It's never compiled into its own dynamic library.

Reply:
Perhaps also worth mentioning, there are actually three copies of `dxcsupport`. One is linked direction into dxc/clang-spirv-tests, one is dynamically linked through `libdxcompiler.so` and a third is loaded at runtime using the `dlopen` and equivalent methods again through `libdxcompiler.so`. That's kind of weird, but that's how the code was originally organized and there is a comment somewhere that says that at least the linking and loading of libdxcompiler.so is intentional to ensure that offsets are the same or something like that. Might be superstition, but I haven't felt confident enough in that to challenge it yet.

Reply:
Cool. I understand this better now. Thanks!

Reply:
Yes, I meant `libdxcompiler.so` `dxcsupport`.